src/witan/models/dem/ccm/components_functions.clj

```diff
@@ -23,6 +25,37 @@
         (ds/rename-columns {col-to-avg avg-name})
         (wds/rollup :mean avg-name [:gss-code :sex :age]))))
 
+(defn jumpoffyr-method-trend
+  "Takes in a dataset with historical data, a column name for a trend to be 
+  calculated for, a name for the trend col and years to start and end to
+  calculate the trend for. Returns a dataset with a new col for the calculated trend"
+  [historical-data trend-col trend-out start-yr-avg end-yr-avg]
+  (let [hist-earliest-yr (reduce min (ds/column historical-data :year))
+        hist-latest-yr (reduce max (ds/column historical-data :year))
+        start-yr (s/validate (s/pred #(>= % hist-earliest-yr)) start-yr-avg)
+        end-yr (s/validate (s/pred #(<= % hist-latest-yr)) end-yr-avg)
+        grouped-data (->> (i/query-dataset historical-data
+                                           {:year {:$gte start-yr
+                                                   :$lte end-yr}})
+                          (i/$group-by [:sex :gss-code :age]))
+        intercept (map (fn [[k v]]
+                         (-> (st/linear-model (i/$ trend-col v) (i/$ :year v))
+                             :coefs
+                             first)) grouped-data)
+        new-dataset (reduce ds/join-rows (map (fn [[k v] i]
```

Comment:
I think the map + reduce means we could replace by a transducer. I'm going to try this in my repl.